### ISSUE TITLE
make folder contents view override optional

### ISSUE THREAD

**Issue by petri** [collective/collective.upload#65]
It's rather annoying that the uploader is always present at the top of folder contents view. Would be better if it was triggered by a marker interface.

**Comment by petri** [comment]
I see that there's a behavior that's marked deprecated. Why not just use the behavior for triggering the folder contents view(let)? If someone wants to have it always on, we can add a separate profile that assigns the behavior on built-in Folder.

**Comment by hvelarde** [comment]
we can also make it configurable via the control panel configlet; I think that's easier.

**Comment by petri** [comment]
Yes, it would be very nice to have it configurable in control panel too. Why not have both? 

I can submit a PR that:

- adds another browser.Folder_Contents registration in configure.zcml, for providers of the behavior
- reuses the behavior as a marker interface
- adds a "foldercontents" profile with an extra layer in browserlayer.xml
- reconfigures the browser.Folder_Contents grok view for that layer

So uploader would be shown in folder contents view of:

1. Any folderish content with behavior set
1. Any folderish content with the marker (behavior interface) set
1. All folderish content, if "foldercontents" profile has been loaded

**Comment by hvelarde** [comment]
I think that could be over-engineered as I see no need to implement this on a content type basis: is better to keep consistency in the site; people could get confused if this is enabled in some folders and disabled in some others.

having a boolean field in the control panel and a condition in the viewlet registration seems enough for this first iteration.

if you can provide a PR we can work in parallel on other features we want to implement to have a new release soon.

**Comment by petri** [comment]
I can see your point. Consistency is good.

However I have a combined www-site/extranet-site -type use case whereby there will be very specific limited areas where always showing the uploader (in folder_contents) is a good idea; it will just clutter the UI elsewhere. Elsewhere, just using the uploader from add menu is sufficient. So that, I need more granular control, either via behavior or maybe even just a marker interface (easy to switch on using collective.interfaces).

I will provide a PR (thought the translations I committed earlier weren't worth a PR).

**Comment by hvelarde** [comment]
as long as I can see all of that involves configuration only and I don't see any reason why that can't be done in a project basis instead of on this package.

I prefer to keep the scope of this package very well delimited and simple; the changes you propose don't add anything especially complex but neither add something of value for the great majority of installations.

@idgserpro @rodfersou @agnogueira comments, please.

**Comment by hvelarde** [comment]
some other points:

* the default will no longer be the same
* we will have 3 different ways to accomplish the same thing
* this will work different for Dexterity and for Archetypes
* unintended consequences if the package is uninstalled and the marker interfaces are still in place

for me hiding the viewlet makes sense on a global basis and what you want to implement can also be solved by using CSS to hide it where you don't want it.

**Comment by petri** [comment]
Thanks for the clear and concise commentary. Yes, the PR is about making the package flexible to configure, both manually and via code.

I may be biased on whether there'd be too much configuration or sufficient value, given the use case and my preference for being able to tweak things also TTW (behaviors & markers) when needed.

The configurability provided by the PR is for developers and integrators rather than site managers, so perhaps it's safe to assume them to know what they are doing? But good points regardless.

Assuming (there'd be) means to globally disable the view (or default to disabled), I guess it would be easy to just register the uploader view on a custom type in ZCML, on a project basis. 

But associating the uploader with content in a more granular or flexible manner would still need something more dynamic AFAIK, and I feel that at point we'd be crossing over from configuring collective.upload to doing something more.

**Comment by idgserpro** [comment]
Is there a way of not customizing folder_contents and just using viewlets, @hvelarde? Or was this customization really needed? We tried to look for the file history about the customization motivations but no luck.

If this folder_contents customization could be changed to a viewlet-only approach, we vote to go that way: anyone can customize the viewlet with it's own needs, in a policy package. The marker interface would be in the policy, and in the render code of the customized viewlet you can check `if not IMyMarkerInterface.providedBy(context):` or something like that to provide this feature of hiding in some situations.

**Comment by petri** [comment]
Thinking about this a bit more, if there'd be control panel setting (using plone.registry) and view condition:

- I can change the PR so it the default can be kept if so desired. In which case the extra profile could be dropped from the PR as well
- Would it then be ok for the view to indiscriminately enable itself if the folder provides a behavior marker interface defined by collective.upload? A policy package could then just enable the behavior or assign the marker.

**Comment by idgserpro** [comment]
@petri Let's wait for an answer from @hvelarde. If it's possible to have everything as viewlets, we would need to remove the folder_contents customization and you can customize this behavior in a custom viewlet your package.

**Comment by hvelarde** [comment]
now I see… you are right: I was thinking the feature was implemented as a viewlet while in fact is an override of the whole template.

we can do the following: remove the whole customization and replace it with a viewlet that can be enabled and disabled using a control panel configlet; after that you can monkey patch the `available` method in any way in your own package to do whatever you want.

the behavior is code that has not been used in years: in the beginning we wanted to implement the functionality that way, but we wouldn't be able to use it in Archetypes folders (all of this code is pre plone.app.contenttypes).

sorry about the mess but this package was one of the first we made and is still plenty of crap.

is probably going to more difficult that expected but we need to be very careful as this is uses in many Government sites and we don't want to break anything neither to change the current behavior.

what do you think?

**Comment by idgserpro** [comment]
If you're going to create new viewlet and don't forget to add an upgradeStep to register it, I don't think we'll run into any issues.

BTW, is a configlet in control panel really necessary if we turn the folder_contents customization into a viewlet? What's the use case for that?

**Comment by hvelarde** [comment]
@idgserpro to enable/disable the upload widget in the folder contents template on a global basis.

**Comment by idgserpro** [comment]
@hvelarde In a global basis, we really don't think it's useful: what's the use of installing collective.upload if you want to disable the upload widget globally?

We still think the best approach is just changing the package to use viewlets, and then anyone can customize the way it wants, using the condition that fits into a particular website workflow.

**Comment by hvelarde** [comment]
@idgserpro you can alway use the **Add new…** menu.

![selection_073](https://cloud.githubusercontent.com/assets/341393/15729264/f78b81f8-2838-11e6-8f98-9f99160a0aaa.png)

**Comment by petri** [comment]
If a viewlet is used, it would be good to bind it to marker interface rather than something all-encompassing such as IFolderish. That would make it easy to selectively enable it on a per-type (behavior assigned) or per-contentitem (marker set) basis, in a policy package.

**Comment by hvelarde** [comment]
@petri you can always override the viewlet with your own behavior too meet your use cases:

http://docs.plone.org/develop/plone/views/more_view_examples.html
